java/src/jmri/jmrix/openlcb/swing/memtool/MemoryToolPane.java

```diff
@@ -0,0 +1,427 @@
+
+        setRunning(false);
+    }
+
+    public MemoryToolPane() {
+    }
+
+    @Override
+    public void dispose() {
+        // and complete this
+        super.dispose();
+    }
+
+    @Override
+    public String getHelpTarget() {
+        return "package.jmri.jmrix.openlcb.swing.memtool.MemoryToolPane";
+    }
+
+    @Override
+    public String getTitle() {
+        if (memo != null) {
+            return (memo.getUserName() + " Memory Tool");
+        }
+        return getTitle(Bundle.getMessage("TitleEventTable"));
+    }
+
+    void pushedCancel(ActionEvent e) {
+        if (running) {
+            cancelled = true;
+        }
+    }
+
+    void setRunning(boolean t) {
+        if (t) {
+            gb.setEnabled(false);
+            pb.setEnabled(false);
+            cb.setEnabled(true);
+        } else {
+            gb.setEnabled(true);
+            pb.setEnabled(true);
+            cb.setEnabled(false);
+        }
+        running = t;
+    }
+
+    int space = 0xFF;
+
+    NodeID farID = new NodeID("0.0.0.0.0.0");
+
+    MemoryConfigurationService.McsReadHandler cbr =
+        new MemoryConfigurationService.McsReadHandler() {
+            @Override
+            public void handleFailure(int errorCode) {
+                setRunning(false);
+                if (errorCode == 0x1082) {
+                    statusField.setText("Done reading");
+                    log.debug("Stopping read due to 0x1082 status");
+                } if (errorCode == 0x1081) {
+                    log.error("Read failed. Address space not known");
+                    statusField.setText("Read failed. Address space not known");
+                } else {
+                    log.error("Read failed. Error code is {}", String.format("%04X", errorCode));
+                    statusField.setText("Read failed. Error code is "+String.format("%04X", errorCode));
+                }
+                try {
+                    outputStream.flush();
+                    outputStream.close();
+                } catch (IOException ex) {
+                    log.error("Error closing file", ex);
+                    statusField.setText("Error closing output file");
+                }
+            }
+
+            @Override
+            public void handleReadData(NodeID dest, int readSpace, long readAddress, byte[] readData) {
+                log.trace("read succeed with {} bytes at {}", readData.length, readAddress);
+                statusField.setText("Read "+readAddress+" bytes");
+                try {
+                    outputStream.write(readData);
+                } catch (IOException ex) {
+                    log.error("Error writing data to file", ex);
+                    statusField.setText("Error writing data to file");
+                    setRunning(false);
+                    return; // stop now
+                }
+                if (readData.length != CHUNKSIZE) {
+                    // short read is another way to indicate end
```

Comment:
The problem is there are some boards that _only_ signal via short read.  The following read just succeeds and returns full-length data.  C.f. that recent vendor discussion.  Maybe I should add another checkbox?

Reply:
On second thought, I think this actually is required behavior.  From the Standard

> Reading at least one byte, but less than the requested amount, due to the length of the address space is not considered an error.
> If the full number of bytes cannot be read for any reason other than the size of the address space, or if no bytes can be read, the Read Reply shall have the Fail bit set and include an error code instead of the requested data.